`r\managed_disk`: Add support for `gallery_image_reference_id`

myc2h6o opened the pull request:
- Fix https://github.com/hashicorp/terraform-provider-azurerm/issues/9890, add a new property `gallery_image_reference_id` to support creating a disk from a gallery image version id
- Test result:
![image](https://user-images.githubusercontent.com/10579712/144786829-9c430297-9b22-4d16-85b6-545e50349c93.png)

myc2h6o commented on internal/services/compute/managed_disk_resource_test.go:
```diff
@@ -758,6 +773,110 @@ resource "azurerm_managed_disk" "test" {
 `, data.Locations.Primary, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
 }
 
+func (r ManagedDiskResource) galleryImageDataDisk(data acceptance.TestData) string {
```
This test does below things:
1. Create a linux vm, with the dependency template from LinuxVirtualMachineResource{}
2. Create a data disk attachment to the vm
3. Create an image gallery, an image definition and an image version from the vm with data disk attached
4. Create the disk out of the data disk of the image in the gallery

katbyte commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -298,12 +306,25 @@ func resourceManagedDiskCreate(d *pluginsdk.ResourceData, meta interface{}) erro
 			ID: utils.String(imageReferenceId),
 		}
+
+		imageReferenceLun := d.Get("image_reference_lun").(int)
+		if imageReferenceLun != -1 {
```
is there a reason we are not using `d.GetOk` ?

myc2h6o replied:
The valid values for `image_reference_lun` are 0, 1, 2, 3, ..., which is the logic unit number of the data disk. In order to make 0 a valid value, I used `-1` as its default value. `d.GetOk` is not able to check the -1 so I used `d.Get` instead.

jackofallops replied:
Hi @myc2h6o - `d.GetOk()` will pick up if the value has been set by the user, so you can remove the `Default` schema property, and this should should allow the value `0`.
Which should let you change this to something like:
```suggestion
		if imageReferenceLun, ok := d.GetOk("image_reference_lun").(int); ok {
```

myc2h6o replied:
I removed the `Default` schema property and used `d.GetOk`, however the `ok` is `false` when I pass `0` in the config. This method seems to [check whether the value is zero](https://github.com/hashicorp/terraform-plugin-sdk/blob/76b3563a7afe40aa29d08d1661c5502b59e3811a/helper/schema/resource_data.go#L88) besides checking if it's set by user.
This property is a little bit tricky. When not present in api, it means os disk, when present, it means index of the data disk and can be set to 0 because the index of data disk starts from 0. In order to include both os disk and data disk, I used -1 to mean os disk, and value >= 0 to mean data disk

tombuildsstuff replied:
What's the use-case for adding this field to Terraform? Presumably this field will be set/conflict when a Disk is added from a virtual machine, so whilst I can see the field exists in the API - I have a feeling exposing this may cause larger issues?

myc2h6o replied:
yes it makes more sense to attach a clean data disk to vm using `azurerm_virtual_machine_data_disk_attachment`. I've removed the disk lun property

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -298,12 +306,25 @@ func resourceManagedDiskCreate(d *pluginsdk.ResourceData, meta interface{}) erro
+			imageDiskReference.Lun = utils.Int32(int32(imageReferenceLun))
+		}
+
+		if _, err := parse.SharedImageVersionID(imageReferenceId); err == nil {
+			log.Printf("[INFO] image_reference_id is a gallery image version id, copying the disk from gallery image.")
+			props.CreationData.GalleryImageReference = imageDiskReference
+		} else {
+			log.Printf("[INFO] image_reference_id is not recognized as a galley image version id, copying the disk from platform/marketplace image.")
+			props.CreationData.ImageReference = imageDiskReference
+		}
```
rather than reusing the existing Property here, we should introduce a new one for this field

myc2h6o replied:
Split it into two properties to make the property meaning clear

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -673,11 +694,13 @@ func resourceManagedDiskRead(d *pluginsdk.ResourceData, meta interface{}) error
-			imageReferenceID := ""
-			if creationData.ImageReference != nil && creationData.ImageReference.ID != nil {
-				imageReferenceID = *creationData.ImageReference.ID
+			if galleryImageReference := creationData.GalleryImageReference; galleryImageReference != nil {
+				setImageReference(d, *galleryImageReference)
+			} else if imageReference := creationData.ImageReference; imageReference != nil {
+				setImageReference(d, *imageReference)
+			} else {
+				setImageReference(d, compute.ImageDiskReference{})
```
there's no need to split this method out, with a separate field for GalleryImageID this becomes simpler and can be inlined, which also moves this left:

```suggestion
			galleryImageId := ""
			imageId := ""
			if galleryImageReference := creationData.GalleryImageReference; galleryImageReference != nil && galleryImageReference.ID != nil {
				galleryImageId = *galleryImageReference.ID
			}
			if imageReference := creationData.ImageReference; imageReference != nil && imageReference.ID != nil {
				imageId = *imageReference.ID
			}
```

myc2h6o replied:
Yeah since imgeReference.Lun is gone, have move the code out of the new method

tombuildsstuff commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -739,3 +762,17 @@ func resourceManagedDiskDelete(d *pluginsdk.ResourceData, meta interface{}) erro
 	return nil
 }
+
+func setImageReference(d *pluginsdk.ResourceData, imageReference compute.ImageDiskReference) {
+	if imageReference.ID != nil {
+		d.Set("image_reference_id", imageReference.ID)
+	} else {
+		d.Set("image_reference_id", "")
+	}
+
+	if imageReference.Lun != nil {
+		d.Set("image_reference_lun", imageReference.Lun)
+	} else {
+		d.Set("image_reference_lun", -1)
+	}
+}
```
which means this can go:

```suggestion
```

myc2h6o replied:
Removed

tombuildsstuff commented on website/docs/r/managed_disk.html.markdown:
```diff
@@ -107,7 +107,9 @@ The following arguments are supported:
-* `image_reference_id` - (Optional) ID of an existing platform/marketplace disk image to copy when `create_option` is `FromImage`.
+* `image_reference_id` - (Optional) ID of an existing platform/marketplace disk image or gallery image version to copy when `create_option` is `FromImage`.
+
+* `image_reference_lun` - (Optional) Index of the data disk in the image when copying from a data disk. Defaults to `-1`, which means to copy from os disk.
```
this default makes no sense in the context of Terraform - `0` is the default LUN for an OS disk in Azure, which we use elsewhere in the Provider - however as mentioned above, perhaps we shouldn't be exposing this field:

```suggestion
```

myc2h6o replied:
Have removed the Lun as there is no meaning to expose it here.
While it seems not correct that `0` is the default Lun for an OS disk. e.g. [virtual_machine_data_disk_attachment_resource](https://github.com/hashicorp/terraform-provider-azurerm/blob/82a4cf8727706199cb5c743a8d86f809c86eed33/internal/services/compute/virtual_machine_data_disk_attachment_resource.go#L55-L60) Lun actually starts with 0, and OS disk doesn't occupy the range

tombuildsstuff reviewed (changes requested):
hey @myc2h6o 

I've taken a look through and left a couple of questions inline about the use-case that we're looking to solve here, since I'm concerned that adding this field will potentially cause issues for other resources here.

Thanks!

myc2h6o commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -673,11 +685,15 @@ func resourceManagedDiskRead(d *pluginsdk.ResourceData, meta interface{}) error
+			imageReferenceId := ""
+			if galleryImageReference := creationData.GalleryImageReference; galleryImageReference != nil && galleryImageReference.ID != nil {
+				galleryImageReferenceId = *galleryImageReference.ID
+			} else if imageReference := creationData.ImageReference; imageReference != nil && imageReference.ID != nil {
```
Use `else if` so that only one of the `image_reference_id` and `gallery_image_reference_id` is assigned.
When galleryImageReferenceId is used, the API assigns the value to imageReferenceId as well, which should be excluded

myc2h6o commented:
Hi @tombuildsstuff, thanks for the suggestions. I've removed the `lun` and split the `gallery_image_reference_id` and `image_reference_id`, please take another look.

katbyte reviewed (approved):
Thanks @myc2h6o - LGTM 🏗️

Weihrauch commented:
Hello,
I don't understand why image_reference_lun parameter has been removed finaly. I have an image in shared image galery which have data disk actually repectively /var /tmp /home mounts. 
How can I create datadisk without selecting the correct disk in image throught lun parameter ? In the current implementation, anytime I try to create managed_disk from image gallery image, it is created from os disk.

myc2h6o commented:
Hi @Weihrauch Earlier I thought it would be better to put everything in OS disk in an image, but as it is a valid use case to support `gallery_image_reference_lun`, I think we may be able to support it as well. In order to give a better summary of this property, I've created a new issue #14959 for discussion and tracking purpose, would you mind subscribing to that issue for update?
